helpbar/helpbar.c

```diff
@@ -0,0 +1,291 @@
+#include "core/lib.h"
+#include "gui/lib.h"
+#include "keymap.h"
+
+static int helpbar_observer(struct NotifyCallback *nc);
+
+/**
+ * make_help - Create one entry for the help bar
+ * @param buf    Buffer for the result
+ * @param buflen Length of buffer
+ * @param txt    Text part, e.g. "delete"
+ * @param menu   Current Menu, e.g. #MENU_PAGER
+ * @param op     Operation, e.g. OP_DELETE
+ * @retval true If the keybinding exists
+ *
+ * This will return something like: "d:delete"
+ */
+static bool make_help(char *buf, size_t buflen, const char *txt, enum MenuType menu, int op)
+{
+  char tmp[128];
+
+  if (km_expand_key(tmp, sizeof(tmp), km_find_func(menu, op)) ||
+      km_expand_key(tmp, sizeof(tmp), km_find_func(MENU_GENERIC, op)))
+  {
+    snprintf(buf, buflen, "%s:%s", tmp, txt);
+    return true;
+  }
+
+  buf[0] = '\0';
+  return false;
+}
+
+/**
+ * compile_help - Create the text for the help menu
+ * @param buf    Buffer for the result
+ * @param buflen Length of buffer
+ * @param menu   Current Menu, e.g. #MENU_PAGER
+ * @param items  Map of functions to display in the help bar
+ * @retval ptr Buffer containing result
+ */
+static char *compile_help(char *buf, size_t buflen, enum MenuType menu,
+                          const struct Mapping *items)
+{
+  char *pbuf = buf;
+
+  for (int i = 0; items[i].name && (buflen > 2); i++)
+  {
+    if (!make_help(pbuf, buflen, _(items[i].name), menu, items[i].value))
+      continue;
+
+    const size_t len = mutt_str_len(pbuf);
+    pbuf += len;
+    buflen -= len;
+
+    *pbuf++ = ' ';
+    *pbuf++ = ' ';
+    buflen -= 2;
+  }
+  return buf;
+}
+
+/**
+ * helpbar_recalc - Recalculate the display of the Help Bar
+ * @param win Help Bar Window
+ *
+ * The Help Bar isn't drawn, yet.
+ */
+static int helpbar_recalc(struct MuttWindow *win)
+{
+  if (!win)
+    return -1;
+
+  struct MuttWindow *win_focus = window_get_focus();
+  if (!win_focus)
+    return 0;
+
+  // Ascend the Window tree until we find help_data
+  while (win_focus && !win_focus->help_data)
+    win_focus = win_focus->parent;
+
+  if (!win_focus)
+    return 0;
+
+  struct HelpbarWindowData *wdata = helpbar_wdata_get(win);
+  if (!wdata)
+    return 0;
+
+  mutt_debug(LL_DEBUG1, "recalc\n");
+
+  char helpstr[1024] = { 0 };
+  compile_help(helpstr, sizeof(helpstr), win_focus->help_menu, win_focus->help_data);
+
+  wdata->help_menu = win_focus->help_menu;
+  wdata->help_data = win_focus->help_data;
+  mutt_str_replace(&wdata->help_str, helpstr);
+  win->actions |= WA_REPAINT;
+
+  return 0;
+}
+
+/**
+ * helpbar_repaint - Redraw the Help Bar
+ * @param win Help Bar Window
+ *
+ * The Help Bar is drawn from the data cached in the HelpbarWindowData.
+ * No calculation is performed.
+ */
+static int helpbar_repaint(struct MuttWindow *win)
+{
+  if (!mutt_window_is_visible(win))
+    return 0;
+
+  struct HelpbarWindowData *wdata = helpbar_wdata_get(win);
+  if (!wdata)
+    return 0;
+
+  mutt_debug(LL_DEBUG1, "repaint\n");
```

Comment:
LL_NOTIFY? Or perhaps you'd like to get rid of all LL_DEBUG1 additions?

Reply:
Ah, my debugging.
Demoted to LL_NOTIFY.